homeassistant/components/heos/services.py

```diff
@@ -71,3 +104,138 @@ async def _sign_out_handler(controller, service):
     except HeosError as err:
         _LOGGER.error("Unable to sign out: %s", err)
+
+
+async def _join_handler(controller, hass, service):
+    """Join a HEOS player to a group."""
+    if controller.connection_state != const.STATE_CONNECTED:
+        _LOGGER.error("Unable to join because HEOS is not connected")
+        return
+    leader = service.data[ATTR_GROUP_LEADER]
+    members = service.data[ATTR_ENTITY_ID]
+
+    try:
+        groups = await controller.get_groups(refresh=True)
+    except HeosError as err:
+        _LOGGER.error("Could not get HEOS group info while joining group. %s", err)
+
+    # Get the HEOS player_id for the leader
+    leader_player_id = None
+    for entity in hass.data[DOMAIN].entities:
+        if isinstance(entity, HeosMediaPlayer) and entity.entity_id == leader:
+            leader_player_id = entity.player_id
+            break
+    if not leader_player_id:
+        _LOGGER.error("Failed to determine HEOS player_id for %s", leader)
+        return
+
+    # Try to find out if the leader entity is already a group leader.
+    # In this case we need to update the existing group.
+    existing_heos_group_id = None
+    existing_heos_group_members = set()
+    for heos_group_id, heos_group in groups.items():
+        if heos_group.leader.player_id == leader_player_id:
+            existing_heos_group_id = heos_group_id
+            existing_heos_group_members = {
+                member.player_id for member in heos_group.members
+            }
+            break
+
+    # Resolve the entities passed in through `members` to HEOS player_ids.
+    new_member_player_ids = {
+        entity.player_id
+        for entity in hass.data[DOMAIN].entities
+        if isinstance(entity, HeosMediaPlayer) and entity.entity_id in members
+    }
+    # In case of an existing HEOS group we also need to include the existing HEOS members to
+    # the group.
+    all_member_player_ids = list(
+        new_member_player_ids.union(existing_heos_group_members)
+    )
+
+    if existing_heos_group_id:
+        _LOGGER.info("Updating HEOS group for %s with members %s", leader, members)
+        try:
+            await controller.update_group(existing_heos_group_id, all_member_player_ids)
+        except HeosError as err:
+            _LOGGER.error(
+                "Unable to call update_group(%s, %s): %s",
+                existing_heos_group_id,
+                all_member_player_ids,
+                err,
+            )
+    else:
+        _LOGGER.info("Creating new HEOS group for %s with members %s", leader, members)
+        try:
+            await controller.create_group(leader_player_id, all_member_player_ids)
+        except HeosError as err:
+            _LOGGER.error(
+                "HEOS create_group with arguments (%s, %s) failed. Error: %s",
+                leader_player_id,
+                members,
+                err,
+            )
+
+
+async def _unjoin_handler(controller, hass, service):
```

Comment:
This is also unnecessarily complex.  `remove_group` also calls `set_group` under the covers.  By passing a single `player_id`, it will ungroup--whether it was a member or the leader.  Simply call this method after resolving the id.

Reply:
Good catch, I have updated the code in `HeosMediaPlayer.async_unjoin()`.